internal/zinc-core/src/main/scala/sbt/internal/inc/RelationBuilder.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Zinc - The incremental compiler for Scala.
+ * Copyright Lightbend, Inc. and Mark Harrah
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package sbt.internal.inc
+
+import sbt.internal.util.Relation
+
+import scala.collection.{ immutable, mutable }
+
+final class RelationBuilder[A <: AnyRef, B <: AnyRef]() {
+  private[this] val forward =
+    new java.util.HashMap[A, (A, mutable.Builder[B, immutable.HashSet[B]])]()
+  private[this] val reverse =
+    new java.util.HashMap[B, (B, mutable.Builder[A, immutable.HashSet[A]])]()
+
+  def update(a: A, b: B): Unit = {
+    val (internedB, asBuilder) =
+      reverse.computeIfAbsent(b, (b => (b, immutable.HashSet.newBuilder[A])))
+    val (internedA, bsBuilder) =
+      forward.computeIfAbsent(a, (a => (a, immutable.HashSet.newBuilder[B])))
+    asBuilder += internedA
+    bsBuilder += internedB
+  }
+
+  def +=(other: Relation[A, B]): Unit = {
+    for ((a, bs) <- other.forwardMap.iterator) {
+      for (b <- bs) {
+        update(a, b)
+      }
+    }
+  }
+
+  def result(): Relation[A, B] = {
+    def toImmutable[K, V](
+        map: java.util.HashMap[K, (K, mutable.Builder[V, immutable.HashSet[V]])]
+    ): immutable.HashMap[K, immutable.HashSet[V]] = {
+      val builder = immutable.HashMap.newBuilder[K, immutable.HashSet[V]]
```

Comment:
https://github.com/scala/scala/pull/8726 is the 2.12.x backport of efficient `immutable.HashMap$Builder` (thanks, @mkeskells!). With that, or with 2.13, this step is step is quite efficient.

I considered an alternative of just doing an `unsafeToImmutableMap` wrapper, but that is a somewhat more fragile if the users of the built `Relation` start to call map operations on it. 

It would be great if we had a richer `immutable.HashMap.Builder` that supported key `getOrElseUpdate` on the map being built, as well as a find `transformValues` step. Then, we could build up the desired structure from the begining and do a second pass that just mutates the data in the leaves.